greenplum-db/gpdb
# Reimplement BRIN indexes for AO/CO tables

This changes the implementation for BRIN indexes on AO/CO tables (please refer to the updated BRIN readme for details).
Please review commits individually.

It also fixes https://github.com/greenplum-db/gpdb/issues/14834 and  https://github.com/greenplum-db/gpdb/issues/15155

Major TODOs:

* Basic performance numbers. See https://github.com/greenplum-db/gpdb/pull/15407#issuecomment-1544338312 [✓]
* Is there any adverse impact to sticking the pagenumber and the next page pointer in the brin special section? [✓]
     - It doesn't change alignment. Having the two extra blknos causes the struct to be 16 bytes, which is fine. It does not break alignment and 16 byte usage is present in other AMs.
* Test cases:
	- **Actually test what we return in the bitmap FROM THE INDEX (as opposed to falling back).** [✓]
	- Whitebox test to assert that we don't create any storage if index is created on an empty table (check if this case is already covered indirectly). [✓]
	- Will always the first logical page number be 1 in the chain? (Currently its probably safe to set it to 1, like we do in positionAtStart as it is reset later on when revmap->currBuf is read...probably need to clean this up and maybe not even set it in positionAtStart...) There will never be any holes in page ranges due to aborted tuples, as summarize/build both fill in the holes. However, we still are opting to read the page number of the first chain member during  positionAtStart to keep things simple and future-proof. [✓]
* Minor:
    - REVMAP_PAGE_MAXITEMS = 5456 could have changed due to our tinkering. It is now 5454. Make appropriate changes to comments etc. [✓]
    -  #define InvalidPageNumber (0) and use that everywhere where we deal with page numbers. [✓]
    - Some more comment updates. [✓]
    - Update to the main chain commit message [✓]
    - CREATE EXTENSION pageinspect in setup.sql so that we can parallelize brin_chain_row and brin_chain_column tests

* Bolierplate:
    - Catalog bump (since we can't bump XLOG_PAGE_MAGIC due to possibility of future merge conflicts) [✓]
    - waldump desc functions need updating [✓]

Out of scope for this PR:
* Refactoring brinbuild to use BlockSequences and loop over them, to be more consistent with other BRIN routines
* Skipping segments pending drop in `relation_get_block_sequences`.
* pageinspect to show logical page numbers of revmap pages and use that in the white-box tests.
* amcheck to check consistency of brin chains
* Evaluate desummarize

Dev-pipeline: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/brin_chain
